crates/cheatnet/src/cheatcodes/spy_events.rs

```diff
@@ -18,33 +20,50 @@ pub enum SpyOn {
     Multiple(Vec<ContractAddress>),
 }
 
+impl SpyOn {
+    pub fn should_be_spied(&mut self, contract_address: ContractAddress) -> bool {
+        match self {
+            SpyOn::All => true,
+            SpyOn::One(address) => *address == contract_address,
+            SpyOn::Multiple(addresses) => addresses.contains(&contract_address),
+        }
+    }
+}
+
 impl CheatnetState {
-    pub fn spy_events(&mut self, spy_on: SpyOn) {
-        self.cheatcode_state.spy_events = Some(spy_on);
+    pub fn spy_events(&mut self, spy_on: SpyOn) -> usize {
+        self.cheatcode_state.spies.push(spy_on);
+        self.cheatcode_state.spies.len() - 1
     }
 
-    pub fn fetch_events(&mut self) -> (usize, Vec<Felt252>) {
+    pub fn fetch_events(&mut self, id: &Felt252) -> (usize, Vec<Felt252>) {
+        let spy_on = &mut self.cheatcode_state.spies[felt_to_usize(id).unwrap()];
```

Comment:
So in case someone passes an id that is invalid, we just want this to panic entirely?

Reply:
Fortunately, users don't have to pass anything here. `EventSpy` structure keeps its id as property and passes it itself